File: src/javascript/algorithms/hashtable/4.groupAnagrams/groupAnagrams.js
function groupAnagramsWithMap(arr) {
    let anagramGroups = new Map();
    for(let str of arr) {
        let sortedString = str.split("").sort().join("");
        if(anagramGroups.has(sortedString)) {
            anagramGroups.get(sortedString).push(str);
        } else {
            anagramGroups.set(sortedString, [str]);
        }
    }
    return Array.from(anagramGroups.values());
}

function groupAnagramsWithObject(arr) {
    let anagramGroups = {};
    for(let str of arr) {
        let sortedString = str.split("").sort().join("");
        let group = anagramGroups[sortedString];
        if(group && group.length > 0) {
            anagramGroups[sortedString].push(str);
        } else {
            anagramGroups[sortedString] = [str];
        }
    }
    return Object.values(anagramGroups);
}

//Without sorting: TC(K * N) SC(K * N)
function groupAnagramsWithoutSorting(arr) {
    let anagramGroups = {};
    for(let str of arr) {

        let charCount = Array(26).fill(0);
        for(const ch of str.toLowerCase()){
            charCount[ch.charCodeAt(0)- 'a'.charCodeAt(0)]
        }
        let charCountkey = charCount.join("");
        if(!anagramGroups[charCountkey]) {
            anagramGroups[charCountkey] = [];
        }
        anagramGroups[charCountkey].push(str);
    }
    return Object.values(anagramGroups);
}

// Multiple Anagrams
console.log("Input: ['eat', 'tea', 'tan', 'ate', 'nat', 'bat']");
console.log("Output1: ", groupAnagramsWithMap(['eat', 'tea', 'tan', 'ate', 'nat', 'bat']));
console.log("Output2: ", groupAnagramsWithoutSorting(['eat', 'tea', 'tan', 'ate', 'nat', 'bat']));
console.log("---------------");

// Mixed Case Anagrams
console.log("Input: ['Eat', 'Tea', 'Tan', 'Ate', 'Nat', 'Bat']");
console.log("Output1: ", groupAnagramsWithObject(['Eat', 'Tea', 'Tan', 'Ate', 'Nat', 'Bat']));
console.log("Output2: ", groupAnagramsWithoutSorting(['Eat', 'Tea', 'Tan', 'Ate', 'Nat', 'Bat']));

console.log("---------------");

// No Anagrams
console.log("Input: ['apple', 'orange', 'banana']");
console.log("Output1: ", groupAnagramsWithMap(['apple', 'orange', 'banana']));
console.log("Output2: ", groupAnagramsWithoutSorting(['apple', 'orange', 'banana']));

console.log("---------------");

File: src/python/algorithms/hashtable/4.groupAnagrams/group_anagrams.py
"""
Group Anagrams

Given an array of strings strs, group the anagrams together.

Time Complexity: O(n * k log k) where k is max string length
Space Complexity: O(n * k)
"""

from collections import defaultdict


def group_anagrams(strs):
    """
    Using sorted string as key.
    TC: O(n * k log k), SC: O(n * k)
    """
    anagram_map = defaultdict(list)
    
    for s in strs:
        key = ''.join(sorted(s))
        anagram_map[key].append(s)
    
    return list(anagram_map.values())


def group_anagrams_count(strs):
    """
    Using character count as key.
    TC: O(n * k), SC: O(n * k)
    """
    anagram_map = defaultdict(list)
    
    for s in strs:
        count = [0] * 26
        for char in s:
            count[ord(char) - ord('a')] += 1
        key = tuple(count)
        anagram_map[key].append(s)
    
    return list(anagram_map.values())


# Test cases
if __name__ == "__main__":
    test_cases = [
        ["eat", "tea", "tan", "ate", "nat", "bat"],
        [""],
        ["a"],
    ]
    
    for strs in test_cases:
        print(f"Input: {strs}")
        print(f"Grouped: {group_anagrams(strs)}")
        print()
